README.md
# synapse-purge [![Docker Hub build][docker-hub-badge]][docker-hub]

> A script to purge content from Synapse servers

This is a simple script to purge content from [Synapse][synapse] servers. It
deletes events, local media and cached remote media up to a specified point in
time. Old user and room avatars are (for now) excluded from getting purged due
to issues with the latest ones getting removed as well. Currently, only Synapse
instances using a [PostgreSQL][postgresql] database are supported.

Note that content on remote (federated) servers can obviously not be deleted
and that Synapse might not consider the most recent content as _history_, which
excludes it from beling deleted even if it is technically inside the specified
purge timeframe.

## Table of contents

+ [Install](#install)
  + [Installing with Docker](#installing-with-docker)
  + [Installing without Docker](#installing-without-docker)
  + [Dependencies](#dependencies)
  + [Updating](#updating)
    + [Updating with Docker](#updating-with-docker)
    + [Updating without Docker](#updating-without-docker)
    + [Upgrading from `1.x.x` to `2.x.x`](#upgrading-from-1xx-to-2xx)
+ [Usage](#usage)
  + [Running with Docker](#running-with-docker)
  + [Running without Docker](#running-without-docker)
  + [Configuration](#configuration)
+ [Maintainer](#maintainer)
+ [Contribute](#contribute)
+ [License](#license)

## Install

The recommended way to run is via [Docker][docker]. Basic instructions on how
to run without it are also provided.

### Installing with Docker

To install for running with Docker, you can simply pull the prebuilt image from
[Docker Hub][docker-hub]:

```zsh
user@local:~$ docker pull mtbl/synapse-purge
```

Alternatively, you can also build the image yourself. The user that is used
inside the container has UID `1000` and GID `1000` by default. You can adjust
this (e.g., to match your host UID/GID) by providing the arguments `USER_ID`
and `GROUP_ID` when making a build.

### Installing without Docker

To install without Docker, you can simply clone the repository and install
dependencies using Poetry.

```zsh
user@local:~$ git clone https://github.com/imtbl/synapse-purge.git
user@local:~$ cd synapse-purge
user@local:synapse-purge$ poetry install
```

### Dependencies

+ [Docker][docker] (when running with Docker)
+ [Python 3.7+][python] (when running without Docker)
+ [Poetry][poetry] (when running without Docker)

### Updating

This script follows [semantic versioning][semantic-versioning] and any breaking
changes that require additional attention will be released under a new major
version (e.g., `2.0.0`). Minor version updates (e.g., `1.1.0` or `1.2.0`) are
therefore always safe to simply install.

When necessary, this section will be expanded with upgrade guides for new major
versions.

#### Updating with Docker

Simply pull the latest Docker image to update:

```zsh
user@local:~$ docker pull mtbl/synapse-purge
```

#### Updating without Docker

If you have installed via cloning the repository, you can update using Git as
well:

```zsh
user@local:synapse-purge$ git pull
```

#### Upgrading from `1.x.x` to `2.x.x`

`2.0.0` has no breaking changes other than the move of the Docker image to a
new location. Please pull `mtbl/synapse-purge` from now on as the old location
is no longer available.

## Usage

### Running with Docker

To make running with Docker as easy as possible, a working
[Docker Compose][docker-compose] example setup is provided. To get started with
this example setup, simply duplicate `docker-compose.yml.example` as
`docker-compose.yml` and adjust the path to the Synapse data directory in the
`volumes` section, the variables in the `environment` section as
described [here](#configuration) and change the `command` if you want to (you
can either run the script once or periodically via cron job).

Take note of the time zone set via the `TZ` environment variable. This is
particularly important for the cron job to run at the time you expect it to.
The time zone has to be set in the
[tz database format][tz-database-time-zones].

Finally, start the container:

```zsh
user@local:synapse-purge$ docker-compose up -d
```

Depending on your choice, the script will now either run once (with command
`run`) or periodically via cron job until stopped (with command `cron`).

The user that is used inside the container when the script is run has the UID
`1000` and the GID `1000` by default. You can change these by providing the
environment variables `CUSTOM_UID` and `CUSTOM_GID` when creating a container.

### Running without Docker

First of all, spawn a shell within the virtual environment (this also creates
the environment if it does not exist yet):

```zsh
user@local:synapse-purge$ poetry shell
```

You can use either environment variables or arguments to configure the script
when running without Docker. The script will look for the provided argument
first and fall back to the respective environment variable if the argument is
not provided.

To see the required and optional arguments, simple launch the script with the
`-h` flag:

```zsh
user@local:synapse-purge$ ./synapse-purge.py -h
usage: synapse-purge.py [-h] [--delta DELTA] [--logging-level {TRACE,DEBUG,INFO,SUCCESS,WARNING,ERROR,CRITICAL}] [--api-url API_URL] [--media-store MEDIA_STORE]
                        [postgres_connection_string] [synapse_auth_token]

positional arguments:
  postgres_connection_string
  synapse_auth_token

optional arguments:
  -h, --help            show this help message and exit
  --delta DELTA, -d DELTA
  --logging-level {TRACE,DEBUG,INFO,SUCCESS,WARNING,ERROR,CRITICAL}, -l {TRACE,DEBUG,INFO,SUCCESS,WARNING,ERROR,CRITICAL}
  --api-url API_URL, -a API_URL
  --media-store MEDIA_STORE, -m MEDIA_STORE
```

### Configuration

Configuration is done via environment variables when running with Docker. These
can also be used when running without Docker, although using launch parameters
instead might be preferable in that case (see [here](#running-without-docker)).

+ `SYNAPSE_PURGE_MEDIA_STORE_PATH=/data/media_store`: the path to the Synapse
  `media_store` directory.
+ `SYNAPSE_PURGE_POSTGRES_CONNECTION_STRING=`:
  the PostgreSQL connection string for the Synapse database, e.g.,
  `host=localhost port=5432 user=synapse dbname=synapse password=password`.
+ `SYNAPSE_PURGE_SYNAPSE_API_URL=http://localhost:8008/`: the base URL of the
  Synapse API. __Trailing `/` required.__
+ `SYNAPSE_PURGE_SYNAPSE_AUTH_TOKEN=`: the access token of an admin user. Can
  be created via `curl -XPOST -d '{"type":"m.login.password", "user":"<userid>", "password":"<password>"}' "http://localhost:8008/_matrix/client/r0/login"`.
+ `SYNAPSE_PURGE_DELTA=86400`: the current date minus the provided value here
  (in seconds) is the date up to which the script will purge.
+ `SYNAPSE_PURGE_LOGGING_LEVEL=INFO`: the desired logging level. Has to be one
  of these: `TRACE`, `DEBUG`, `INFO`, `SUCCESS`, `WARNING`, `ERROR`, `CRITICAL`.
+ `SYNAPSE_PURGE_DOCKER_CRON_SCHEDULE=0 4 * * *`: the cron schedule for running
  with Docker using the `cron` command.

## Maintainer

[imtbl][maintainer]

## Contribute

You are welcome to help out!

[Open an issue][issues] or submit a pull request.

## License

[AGPLv3](LICENSE) © imtbl

[synapse]: https://github.com/matrix-org/synapse/
[postgresql]: https://www.postgresql.org/
[docker]: https://www.docker.com/
[docker-hub]: https://hub.docker.com/r/mtbl/synapse-purge/
[python]: https://www.python.org/
[poetry]: https://python-poetry.org/
[semantic-versioning]: https://semver.org/
[docker-compose]: https://docs.docker.com/compose/
[tz-database-time-zones]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

[docker-hub-badge]: https://img.shields.io/docker/cloud/automated/mtbl/synapse-purge.svg

[maintainer]: https://github.com/imtbl
[issues]: https://github.com/imtbl/synapse-purge/issues/new


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.2.0] - 2020-10-23

### Changed

+ Maintenance release, no feature changes

## [2.1.0] - 2020-07-26

### Fixed

+ Made it so missing `purge_id` fields do not cause synapse-purge to error and
  exit
+ Made it so `m.room.avatar` events with missing `url` fields do not cause
  synapse-purge to error and exit

## [2.0.0] - 2020-07-01

### Changed

+ Added media thumbnail purging
+ Clarified that the printed timestamps are UTC

## [1.5.0] - 2020-01-29

### Added

+ Added the ability to install dependencies using Poetry
+ Added automatic code formatting using Black

## [1.4.1] - 2020-01-12

### Fixed

+ Fixed the usage of `USER_ID` instead of `GROUP_ID` in Docker image builds

## [1.4.0] - 2020-01-12

### Removed

+ Removed the ability to define custom UID and GID for the Docker container at
  container creation due to several issues arising from that

## [1.3.0] - 2020-01-12

### Changed

+ Changed the default Docker container behavior to only run synapse-purge once
  instead of starting a cron job

## [1.2.0] - 2020-01-07

### Added

+ Added the ability to define custom UID and GID for the Docker container at
  container creation

### Changed

+ Updated dependencies

## [1.1.0] - 2019-12-07

### Changed

+ Old user and room avatars are now excluded from getting purged due to issues
  with the latest ones getting removed as well

## 1.0.0 - 2019-11-04

### Added

+ Initial release

[Unreleased]: https://github.com/imtbl/synapse-purge/compare/2.2.0...develop
[2.2.0]: https://github.com/imtbl/synapse-purge/compare/2.1.0...2.2.0
[2.1.0]: https://github.com/imtbl/synapse-purge/compare/2.0.0...2.1.0
[2.0.0]: https://github.com/imtbl/synapse-purge/compare/1.5.0...2.0.0
[1.5.0]: https://github.com/imtbl/synapse-purge/compare/1.4.1...1.5.0
[1.4.1]: https://github.com/imtbl/synapse-purge/compare/1.4.0...1.4.1
[1.4.0]: https://github.com/imtbl/synapse-purge/compare/1.3.0...1.4.0
[1.3.0]: https://github.com/imtbl/synapse-purge/compare/1.2.0...1.3.0
[1.2.0]: https://github.com/imtbl/synapse-purge/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/imtbl/synapse-purge/compare/1.0.0...1.1.0


synapse-purge.py
#!/usr/bin/env python

# synapse-purge
# Copyright (C) 2019-present  Chris Braun  https://github.com/cryzed
# Copyright (C) 2019-present  imtbl  https://github.com/imtbl

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import enum
import json
import os
import sys
import time
import urllib.parse
from datetime import datetime, timedelta
from typing import Optional, Set, List, Dict

import postgres
import psycopg2
import requests
from loguru import logger

PURGE_HISTORY_API_ENDPOINT = "/_synapse/admin/v1/purge_history/{room}"
PURGE_HISTORY_STATUS_API_ENDPOINT = "/_synapse/admin/v1/purge_history_status/{id}"
PURGE_REMOTE_MEDIA_API_ENDPOINT = "/_synapse/admin/v1/purge_media_cache"


class ExitCode(enum.IntEnum):
    Success = 0
    Failure = 1


class PurgeStatus(str, enum.Enum):
    Active = "active"
    Complete = "complete"
    Failed = "failed"


def get_argument_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "postgres_connection_string",
        nargs="?",
        default=os.getenv("SYNAPSE_PURGE_POSTGRES_CONNECTION_STRING"),
    )
    parser.add_argument(
        "synapse_auth_token",
        nargs="?",
        default=os.getenv("SYNAPSE_PURGE_SYNAPSE_AUTH_TOKEN"),
    )
    parser.add_argument(
        "--delta",
        "-d",
        type=int,
        default=os.getenv("SYNAPSE_PURGE_DELTA", 60 * 60 * 24),
    )
    parser.add_argument(
        "--logging-level",
        "-l",
        choices={"TRACE", "DEBUG", "INFO", "SUCCESS", "WARNING", "ERROR", "CRITICAL"},
        default=os.getenv("SYNAPSE_PURGE_LOGGING_LEVEL", "INFO"),
    )
    parser.add_argument(
        "--api-url",
        "-a",
        default=os.getenv("SYNAPSE_PURGE_SYNAPSE_API_URL", "http://localhost:8008/"),
    )
    parser.add_argument(
        "--media-store",
        "-m",
        default=os.getenv("SYNAPSE_PURGE_MEDIA_STORE_PATH", "/data/media_store"),
    )
    return parser


def get_api_url(base_url: str, suffix: str) -> str:
    return urllib.parse.urljoin(base_url, suffix)


def get_delta_date(seconds: int) -> datetime:
    return datetime.utcnow() - timedelta(seconds=seconds)


def get_last_event_id(
    db: postgres.Postgres, room_id: str, before: datetime
) -> Optional[str]:
    return db.one(
        "SELECT event_id FROM events WHERE received_ts <= %(before)s AND room_id=%(room_id)s ORDER BY received_ts DESC LIMIT 1",
        before=int(before.timestamp() * 1000),
        room_id=room_id,
    )


def get_important_media_ids(db: postgres.Postgres) -> Set[str]:
    media_ids = set()

    for event in db.all("SELECT json FROM event_json"):
        data = json.loads(event)
        content = data["content"]
        type_ = data["type"]

        if type_ == "m.room.member":
            avatar_url = content.get("avatar_url")
            if avatar_url:
                media_ids.add(avatar_url)
        elif type_ == "m.room.avatar":
            if "url" not in content:
                logger.warning("No URL defined for {!r}-event: {!r}", type_, data)
                continue

            media_ids.add(content["url"])

    return set(urllib.parse.urlsplit(url).path[1:] for url in media_ids)


def get_local_media_record_ids(db: postgres.Postgres, before: datetime) -> List[str]:
    return db.all(
        "SELECT media_id FROM local_media_repository WHERE created_ts <= %(before)s",
        before=int(before.timestamp() * 1000),
    )


def get_room_record_ids(db: postgres.Postgres) -> List[str]:
    return db.all("SELECT room_id FROM rooms")


def delete_local_media_record(db: postgres.Postgres, media_id: str) -> None:
    db.run(
        "DELETE FROM local_media_repository WHERE media_id = %(media_id)s",
        media_id=media_id,
    )


def purge_history(
    session: requests.Session, api_url: str, room_id: str, event_id: str
) -> Optional[str]:
    url = get_api_url(api_url, PURGE_HISTORY_API_ENDPOINT).format(room=room_id)
    payload = {"delete_local_events": True, "purge_up_to_event_id": event_id}
    return session.post(url, json=payload).json().get("purge_id")


def purge_history_status(
    session: requests.Session, api_url: str, purge_id: str
) -> PurgeStatus:
    url = get_api_url(api_url, PURGE_HISTORY_STATUS_API_ENDPOINT).format(id=purge_id)
    return PurgeStatus(session.get(url).json()["status"])


def purge_remote_media(
    session: requests.Session, api_url: str, before: datetime
) -> Dict:
    url = get_api_url(api_url, PURGE_REMOTE_MEDIA_API_ENDPOINT)
    return session.post(
        url, params={"before_ts": int(before.timestamp() * 1000)}
    ).json()


def get_local_media_paths(media_store_path: str, media_id: str) -> List[str]:
    paths = []
    for folder_name in "local_content", "local_thumbnails":
        path = os.path.join(
            media_store_path, folder_name, media_id[:2], media_id[2:4], media_id[4:]
        )
        paths.append(path)

    return paths


def wait_for_purge(
    session: requests.Session, api_url: str, purge_id: str, interval: float = 1
) -> PurgeStatus:
    while True:
        status = purge_history_status(session, api_url, purge_id)
        if status != PurgeStatus.Active:
            return status

        time.sleep(interval)


def main(arguments: argparse.Namespace) -> ExitCode:
    logger.remove()
    logger.add(sys.stderr, level=arguments.logging_level)
    logger.debug("Called with {!r}", arguments)

    if not (arguments.postgres_connection_string and arguments.synapse_auth_token):
        logger.error(
            "Postgres connection string and Synapse auth token can not be empty"
        )
        return ExitCode.Failure

    try:
        db = postgres.Postgres(arguments.postgres_connection_string)
    except (psycopg2.OperationalError, psycopg2.InternalError):
        logger.exception(
            "Connecting to database using {!r} failed:",
            arguments.postgres_connection_string,
        )
        return ExitCode.Failure

    session = requests.session()
    session.headers["Authorization"] = "Bearer " + arguments.synapse_auth_token

    before_date = get_delta_date(int(arguments.delta))
    before_date_string = before_date.isoformat(sep=" ", timespec="seconds")
    logger.info("Purging events up to: {} UTC", before_date_string)

    # Purge room history for all rooms
    rooms = get_room_record_ids(db)
    room_count = len(rooms)
    for index, room_id in enumerate(rooms, start=1):
        logger.info("({}/{}) Processing room: {!r}...", index, room_count, room_id)
        event_id = get_last_event_id(db, room_id, before_date)
        if event_id is None:
            logger.info(
                "No event ID before: {} UTC for room: {!r}, skipping",
                before_date_string,
                room_id,
            )
            continue

        logger.info(
            "Last event ID before: {} UTC for room {!r}: {!r}",
            before_date_string,
            room_id,
            event_id,
        )
        purge_id = purge_history(session, arguments.api_url, room_id, event_id)
        if not purge_id:
            logger.warning("Failed to purge room: {!r}: received no purge ID", room_id)
            continue

        logger.info("Purging room: {!r} in progress: {!r}...", room_id, purge_id)
        result = wait_for_purge(session, arguments.api_url, purge_id)
        logger.info("Purged room: {!r} with status: {!r}", room_id, result)

    logger.info("Purging local media older than: {} UTC...", before_date_string)
    local_media = get_local_media_record_ids(db, before_date)
    important_files = get_important_media_ids(db)

    # Purge local media manually
    old_media = set(local_media) - important_files
    old_media_count = len(old_media)
    logger.info("{} media to be cleaned...", len(old_media))
    for index, media_id in enumerate(old_media, start=1):
        logger.info(
            "({}/{}) Processing media: {!r}...", index, old_media_count, media_id
        )
        paths = get_local_media_paths(arguments.media_store, media_id)
        for path in paths:
            if not os.path.isfile(path):
                logger.debug("{!r} could not be found or is not a file", path)
                continue
            os.remove(path)

        delete_local_media_record(db, media_id)

    # Purge remote media
    logger.info("Purging cached remote media older than: {} UTC...", before_date_string)
    result = purge_remote_media(session, arguments.api_url, before_date)
    logger.info("Purged cached remote media: {!r}", result)
    return ExitCode.Success


if __name__ == "__main__":
    parser = get_argument_parser()
    arguments = parser.parse_args()
    parser.exit(main(arguments))
